src/pypyjs.js

```diff
@@ -0,0 +1,1103 @@
+  js.globals['pypyjs']._lastErrorTrace = err_trace
+`;
+
+    const code_chars = Module.intArrayFromString(_code);
+    code_ptr = Module.allocate(code_chars, 'i8', Module.ALLOC_NORMAL);
+    if (!code_ptr) {
+      throw new pypyjs.Error('Failed to allocate memory');
+    }
+
+    const res = Module._pypy_execute_source(code_ptr);
+    if (res < 0) {
+      throw new pypyjs.Error('Error executing python code');
+    }
+
+    Module._free(code_ptr);
+
+    // XXX TODO: races/re-entrancy on _lastError?
+    if (pypyjs._lastErrorName) {
+      const err = new pypyjs.Error(
+        pypyjs._lastErrorName,
+        pypyjs._lastErrorMessage,
+        pypyjs._lastErrorTrace
+      );
+      pypyjs._lastErrorName = null;
+      pypyjs._lastErrorMessage = null;
+      pypyjs._lastErrorTrace = null;
+      reject(err);
+    }
+
+    resolve(null);
+  });
+};
+
+// Method to determine when the interpreter is ready.
+//
+// This method returns a promise that will resolve once the interpreter
+// is ready for use.
+//
+pypyjs.prototype.ready = function ready() {
+  return this._ready;
+};
+
+// Method to execute some python code.
+//
+// This passes the given python code to the VM for execution.
+// It's fairly directly analogous to the 'exec" statement in python.
+// It is not possible to directly access the result of the code, if any.
+// Rather you should store it into a variable and then use the get() method.
+//
+pypyjs.prototype.exec = function exec(code, options) {
+  return this._ready.then(() => {
+    let promise = Promise.resolve();
+    let preCode;
+
+    // Find any "import" statements in the code,
+    // and ensure the modules are ready for loading.
+    if (this.autoLoadModules) {
+      promise = promise.then(() => {
+        return this.findImportedNames(code);
+      })
+      .then((imports) => {
+        return this.loadModuleData.apply(this, imports);
+      });
+    }
+
+    // if any modules have been re-added then we need to remove them from
+    // sys.modules which clears them from memory and allows them to be reloaded
+    // from the emscripten file system
+    if (Object.keys(this._modulesToReset).length) {
+      // construct python code to remove module from sys.modules:
+      // ```python
+      //   import sys
+      //   if 'foo' in sys.modules: del(sys.modules['foo'])
+      // ```
+      const modulesToLoad =
+        Object.keys(this._modulesToReset)
+          .map(mod => `if '${mod}' in sys.modules: del(sys.modules['${mod}'])`);
+
+      preCode = `try:
+  import sys
+  ${modulesToLoad.join('\n  ')}
+except:
+  raise SystemError('Failed to reload custom modules')`;
+      this._modulesToReset = {};
+    }
+
+    // Now we can execute the code in custom top-level scope.
+    const _code = (options && options.file)
+      ? `top_level_scope['__file__'] = '${options.file}'; execfile('${options.file}', top_level_scope.__dict__)`
+      : `exec '''${_escape(code)}''' in top_level_scope.__dict__`;
```

Comment:
If you pass in `options.file`, does that mean the `code` argument is ignored?

More broadly, it's not obvious to me why `pypyjs.prototype.execfile` was refactored to pass in additional options to this method, can you say a little about what's behind that change?